lib/xcp_const.ml

```diff
@@ -0,0 +1 @@
+let good_ciphersuites = "!SSLv2:!EXPORT:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-GCM-SHA384:AES256-SHA256:AES128-SHA256:-MD5:-SSLv3:-RC4"
```

Comment:
Why `-SSLv3`, but `!SSLv2`? The docs say that `!` permanently removes it from the list in a way that it cannot be added back even if explicitly stated. I don't think we want to add these back, so it'd be better to specify things with `!` at the beginning.
It would perhaps be better to say:
`!SSLv2:!SSLv3:!EXPORT:!MD5:!RC4:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-GCM-SHA384:AES256-SHA256:AES128-SHA256`

`openssl ciphers -s -tls1_2 -v '!SSLv2:!SSLv3:!EXPORT:!MD5:!RC4:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-GCM-SHA384:AES256-SHA256:AES128-SHA256'` says:
```
ECDHE-RSA-AES256-SHA384 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AES(256)  Mac=SHA384
ECDHE-RSA-AES256-GCM-SHA384 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AESGCM(256) Mac=AEAD
AES256-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(256)  Mac=SHA256
AES128-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(128)  Mac=SHA2
```

In fact since we specify our cipherlist with explicit cipher suite names I don't think we necessarily need all the `!` entries, but I think it is good practice so we don't (or OpenSSL doesn't) accidentally enable something we definitely don't want.

I would also add `!SHA1`, it is not in our ciphersuite currently, and it is getting disabled in browsers as well.

Reply:
I don't think removing the ssl/tls ciphersuites will do what we think they do.

> TLSv1.2, TLSv1.0, SSLv3
>    Lists ciphersuites which are only supported in at least TLS v1.2, TLS v1.0 or SSL v3.0 respectively. Note: there are no ciphersuites specific to TLS v1.1. Since this is only the minimum version, if, for example, TLSv1.0 is negotiated then both TLSv1.0 and SSLv3.0 ciphersuites are available.
>    Note: these cipher strings do not change the negotiated version of SSL or TLS, they only affect the list of available cipher suites.

Removing them doesn't meant only unsecure ciphersuites get removed, but other may be removed as well.

Not disallowing them doesn't seem to have an immediate effect on 8.1 (build from late Jan):

```
# openssl ciphers -v '!EXPORT:!MD5:!RC4:!SHA1:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-GCM-SHA384:AES256-SHA256:AES128-SHA256'
ECDHE-RSA-AES256-SHA384 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AES(256)  Mac=SHA384
ECDHE-RSA-AES256-GCM-SHA384 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AESGCM(256) Mac=AEAD
AES256-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(256)  Mac=SHA256
AES128-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(128)  Mac=SHA256


# openssl ciphers -v '!SSLv2:!SSLv3:!TLSv1.0:!EXPORT:!MD5:!RC4:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-GCM-SHA384:AES256-SHA256:AES128-SHA256'
ECDHE-RSA-AES256-SHA384 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AES(256)  Mac=SHA384
ECDHE-RSA-AES256-GCM-SHA384 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AESGCM(256) Mac=AEAD
AES256-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(256)  Mac=SHA256
AES128-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(128)  Mac=SHA256
```

Reply:
Then perhaps just use the explicit ciphersuite names for simplicity? We do not use things like HIGH/LOW/ALL/DEFAULT/etc. that would expand to multiple ciphersuites, we already know exactly the ciphersuites we want, the only reason to have the exclusions is to prevent us accidentally adding a ciphersuite we don't want, but since this list is fairly short the risk of that is fairly low.